Fix schema nullability for datastores written and read using avro format

It appears that saving avro datastore using `AvroDataFrameWriter` and subsequent reading using `AvroDataFrameReader` fails  due to wrong nullability of schema fields:
```shell-script
scala>     val loaded = new AvroDataFrameReader(spark).read("/data/elsevier/documenttext/2021-05-17", DocumentText.SCHEMA$)
loaded: org.apache.spark.sql.DataFrame = [id: string, text: string]

scala>     loaded.printSchema()
root
 |-- id: string (nullable = true)
 |-- text: string (nullable = true)

scala>     new AvroDataFrameWriter(loaded).write("/tmp/przemyslaw.jacewicz/tmp", DocumentText.SCHEMA$)
21/05/19 10:27:50 WARN scheduler.TaskSetManager: Lost task 24.0 in stage 1.0 (TID 49, eos-m2-sn11.ocean.icm.edu.pl, executor 4): org.apache.spark.SparkException: Task failed while writing rows.
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask(FileFormatWriter.scala:257)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:170)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:169)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1405)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.avro.AvroRuntimeException: Not a union: "string"
	at org.apache.avro.Schema.getTypes(Schema.java:298)
	... org.apache.spark.sql.avro.AvroSerializer.org$apache$spark$sql$avro$AvroSerializer$$resolveNullableType(AvroSerializer.scala:229)
	...
```
We should expect that writing and reading dataframes using IIS support for avro dataframes should properly handle nullability of fields. 

We should also check if this problem appears for `AvroDatasetReader` and `AvroDatasetWriter`.

--- a/iis-common/src/main/scala/eu/dnetlib/iis/common/spark/avro/AvroDataFrameReader.scala
+++ b/iis-common/src/main/scala/eu/dnetlib/iis/common/spark/avro/AvroDataFrameReader.scala
@@ -31,9 +31,10 @@ class AvroDataFrameReader(val spark: SparkSession) extends Serializable {
    * @return DataFrame with data read from given path.
    */
   def read(path: String, avroSchema: Schema): DataFrame = {
-    spark.read
+    val in = spark.read
       .format("avro")
       .option("avroSchema", avroSchema.toString)
       .load(path)
+    spark.createDataFrame(in.rdd, SchemaConverters.toSqlType(avroSchema).dataType.asInstanceOf[StructType])
   }
 }

--- a/iis-common/src/main/scala/eu/dnetlib/iis/common/spark/avro/AvroDatasetReader.scala
+++ b/iis-common/src/main/scala/eu/dnetlib/iis/common/spark/avro/AvroDatasetReader.scala
@@ -1,9 +1,8 @@
 package eu.dnetlib.iis.common.spark.avro
 
-import com.fasterxml.jackson.databind.{DeserializationFeature, ObjectMapper}
 import org.apache.avro.Schema
 import org.apache.avro.specific.SpecificRecordBase
-import org.apache.spark.sql.{Dataset, Encoder, Encoders, SparkSession}
+import org.apache.spark.sql.{Dataset, SparkSession}
 
 /**
  * Support for reading avro datastores as datasets.
@@ -25,13 +24,6 @@ class AvroDatasetReader(val spark: SparkSession) extends Serializable {
    * @return Dataset with data read from given path.
    */
   def read[T <: SpecificRecordBase](path: String, avroSchema: Schema, clazz: Class[T]): Dataset[T] = {
-    implicit val encoder: Encoder[T] = Encoders.kryo(clazz)
-    val mapper = new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
-    spark.read
-      .format("avro")
-      .option("avroSchema", avroSchema.toString)
-      .load(path)
-      .toJSON
-      .map(json => mapper.readValue(json, clazz))
+    new AvroDataFrameSupport(spark).toDS(new AvroDataFrameReader(spark).read(path, avroSchema), clazz)
   }
 }

--- a/iis-common/src/test/java/eu/dnetlib/iis/common/spark/avro/AvroDataFrameReaderTest.java
+++ b/iis-common/src/test/java/eu/dnetlib/iis/common/spark/avro/AvroDataFrameReaderTest.java
@@ -3,7 +3,6 @@ package eu.dnetlib.iis.common.spark.avro;
 import eu.dnetlib.iis.common.avro.Person;
 import eu.dnetlib.iis.common.spark.TestWithSharedSparkSession;
 import eu.dnetlib.iis.common.utils.AvroTestUtils;
-import org.apache.avro.Schema;
 import org.apache.spark.sql.Dataset;
 import org.apache.spark.sql.Row;
 import org.apache.spark.sql.avro.SchemaConverters;
@@ -40,7 +39,7 @@ class AvroDataFrameReaderTest extends TestWithSharedSparkSession {
         Dataset<Row> result = reader.read(inputDir.toString(),
                 (StructType) SchemaConverters.toSqlType(Person.SCHEMA$).dataType());
 
-        assertSchemasEqualIgnoringNullability(Person.SCHEMA$, result.schema());
+        assertEquals(SchemaConverters.toSqlType(Person.SCHEMA$).dataType(), result.schema());
         List<Row> rows = result.collectAsList();
         assertEquals(1, rows.size());
         Row row = rows.get(0);
@@ -58,16 +57,12 @@ class AvroDataFrameReaderTest extends TestWithSharedSparkSession {
 
         Dataset<Row> result = reader.read(inputDir.toString(), Person.SCHEMA$);
 
-        assertSchemasEqualIgnoringNullability(Person.SCHEMA$, result.schema());
+        assertEquals(SchemaConverters.toSqlType(Person.SCHEMA$).dataType(), result.schema());
         List<Row> rows = result.collectAsList();
         assertEquals(1, rows.size());
         Row row = rows.get(0);
         assertEquals(person.getId(), row.getAs("id"));
         assertEquals(person.getName(), row.getAs("name"));
         assertEquals(person.getAge(), row.getAs("age"));
-    }
-
-    private static void assertSchemasEqualIgnoringNullability(Schema avroSchema, StructType sqlSchema) {
-        assertEquals(SchemaConverters.toSqlType(avroSchema).dataType().asNullable(), sqlSchema.asNullable());
     }
 }

